### ISSUE TITLE
Support for multiple accounts

### ISSUE THREAD

**Issue by AArnott** [zcash/librustzcash#1335]
It seems we need a high-level issue tracking overall support for multiple accounts:

Here are some work items that occur to me:
- `WalletWrite` should expose a function that allows adding HD-derived accounts at a particular index. The existing `create_account` function assumes "last index + 1" which is a great policy, but doesn't provide the app with even an awareness of which index was created. I feel the app should also be able to control this, but with fn-level docs urging the app to only create one account index at a time and in-order. (see also #628)
- The ability to import accounts via UFVK and UIVK. A parameter may indicate whether the caller expects that a spending key is available for this account so that sync knows whether to maintain commitment trees (or whatever the marginal cost is for spendable accounts).
- Adding a second account should trigger partial re-scan. That is, the `scan_queue` should indicate that blocks must be re-scanned. Ideally, the rescan would _only_ run trial decryptions for recently added accounts that have not yet been scanned.
- Removing accounts

**Comment by str4d** [comment]
In a meeting today, @daira @nuttycom and I discussed how wallet apps should identify accounts. We discussed several options, including deriving a persistent account-level identifier from the seed phrase, or deriving something from the Orchard IVK, but each of these ran into problems with various use cases (e.g. imporing the ZecPages viewing key, which has no Orchard component and for which we couldn't get a different fingerprint).

Instead, we've decided to continue with account identifiers that are local-only: they are only valid in the context of a specific wallet database, and MUST NOT be cached by wallet apps across wallet recreation events. Our plan is to:

- `zcash_client_backend`:
  - Augment `WalletRead::AccountId` with the ability to encode to, and parse from, opaque bytes. This encoding MUST NOT be parsed or otherwise interpreted by wallet apps.
  - Document on `AccountId` its local-only properties and how wallet apps must treat them.
- `zcash_client_sqlite` specifically:
  - Introduce a per-wallet UUID (in its own database as a singleton, that is generated via a migration)
  - Encode its account IDs (which are currently the account's primary key from the `accounts` table) as the tuple `(wallet_uuid, account_primary_key)`.

**Comment by str4d** [comment]
> This doesn't need to be something that's exposed at the `zcash_client_backend` level

Not having encoding functions at the `WalletRead` level makes for potentially inconsistent UX (hard to tell given that we only have a single impl in the repo). But I agree that given the caching / encoding here is at the wallet app level, the `WalletRead` implementation is concretely fixed. So for now I agree we can just expose this from `zcash_client_sqlite`; updated the list.

> > Encode its account IDs (which are currently the account's primary key from the accounts table) as the tuple (wallet_uuid, account_primary_key).
> 
> An alternative that we discussed to this was to assign a distinct UUID to each account, and map that to an account ID prior to performing queries. After some consideration, I think that I prefer this solution to the tuple-based approach, because when we add a persistent database reference, we can just have an in-memory map that acts as a cache of the `(account_uuid, account_primary_key)` mappings.
> 
> This also admits a client-side solution: just use the primary-key-based account identifier across the FFI, but add a `zcash_client_sqlite` method that the wallet can use to retrieve and cache those mappings.

In the meeting, the rationale we had for using the tuple was that it meant you didn't need to perform that mapping lookup to get the account primary key (you instead just had to validate the per-wallet UUID), and that existing downstreams that were using the account primary key as the account ID had a potentially easier migration path. But if we're willing to do lookups from `account_uuid` to `account_primary_key` (as we already do with lookups from UFVK / UIVK to `account_id`) then that's fine. However then we need to have two separate types in the crate's public API (the current `AccountId`, and something else), with a method for converting between them.

**Comment by nuttycom** [comment]
> However then we need to have two separate types in the crate's public API (the current AccountId, and something else), with a method for converting between them.

Right, essentially what I was thinking of is something along the lines of `AccountUuid`, and then having a method that returns `BTreeMap<AccountUuid, AccountId>`.

**Comment by AArnott** [comment]
While the value can be opaque, it *should* be serializable. My SDK client keeps its own database alongside the sqlite one from the SDK, and my DB needs to be able to refer to accounts in the sqlite db. So whatever it is, I need a durable identifier. The sqlite db has one perfectly handy, compact and efficient at its disposal that it can share. Using a UFVK or some other substitute substantially complicates things and potentially makes them significantly slower.

**Comment by str4d** [comment]
#1631 implements the UUID-per-account approach.
